File: apps/elixir/lib/algorithm/stacks_and_queues.ex
defmodule Algorithm.Stack do
  @moduledoc "Agent ベースのスタック"

  def new, do: Agent.start_link(fn -> [] end)

  def push(pid, value), do: Agent.update(pid, fn s -> [value | s] end)

  def pop(pid) do
    Agent.get_and_update(pid, fn
      [] -> {:empty, []}
      [h | t] -> {{:ok, h}, t}
    end)
  end

  def peek(pid) do
    Agent.get(pid, fn
      [] -> :empty
      [h | _] -> {:ok, h}
    end)
  end

  def empty?(pid), do: Agent.get(pid, &(&1 == []))
end

defmodule Algorithm.Queue do
  @moduledoc "Agent ベースのキュー（front/rear リスト方式）"

  def new, do: Agent.start_link(fn -> {[], []} end)

  def enqueue(pid, value),
    do: Agent.update(pid, fn {front, rear} -> {front, [value | rear]} end)

  def dequeue(pid) do
    Agent.get_and_update(pid, fn
      {[], []} ->
        {:empty, {[], []}}

      {[], rear} ->
        [h | t] = Enum.reverse(rear)
        {{:ok, h}, {t, []}}

      {[h | t], rear} ->
        {{:ok, h}, {t, rear}}
    end)
  end

  def empty?(pid), do: Agent.get(pid, fn {f, r} -> f == [] and r == [] end)
end

File: apps/clojure/src/algorithm/stacks_and_queues.clj
(ns algorithm.stacks-and-queues
  "第4章 スタックとキュー")

;; --- スタック ---

(defn make-stack
  "固定長スタックを作成する"
  [capacity]
  (atom {:data [] :max capacity}))

(defn stack-empty?
  "スタックが空かどうかを判定する"
  [s]
  (empty? (:data @s)))

(defn stack-full?
  "スタックが満杯かどうかを判定する"
  [s]
  (>= (count (:data @s)) (:max @s)))

(defn stack-push
  "スタックに値をプッシュする"
  [s val]
  (if (stack-full? s)
    (throw (Exception. "Stack is full"))
    (swap! s update :data conj val)))

(defn stack-pop
  "スタックから値をポップする"
  [s]
  (if (stack-empty? s)
    (throw (Exception. "Stack is empty"))
    (let [v (peek (:data @s))]
      (swap! s update :data pop)
      v)))

(defn stack-peek
  "スタックの先頭要素を参照する（取り出さない）"
  [s]
  (if (stack-empty? s)
    (throw (Exception. "Stack is empty"))
    (peek (:data @s))))

;; --- キュー（リングバッファ） ---

(defn make-queue
  "固定長キューを作成する"
  [capacity]
  (atom {:data (vec (repeat capacity nil))
         :capacity capacity
         :front 0
         :rear 0
         :num 0}))

(defn queue-empty?
  "キューが空かどうかを判定する"
  [q]
  (zero? (:num @q)))

(defn queue-full?
  "キューが満杯かどうかを判定する"
  [q]
  (>= (:num @q) (:capacity @q)))

(defn enqueue
  "キューに値をエンキューする"
  [q val]
  (if (queue-full? q)
    (throw (Exception. "Queue is full"))
    (let [{:keys [rear capacity]} @q]
      (swap! q (fn [state]
                 (-> state
                     (assoc-in [:data rear] val)
                     (assoc :rear (mod (inc rear) capacity))
                     (update :num inc)))))))

(defn dequeue
  "キューから値をデキューする"
  [q]
  (if (queue-empty? q)
    (throw (Exception. "Queue is empty"))
    (let [{:keys [front capacity data]} @q
          v (nth data front)]
      (swap! q (fn [state]
                 (-> state
                     (assoc :front (mod (inc front) capacity))
                     (update :num dec))))
      v)))

(defn queue-peek
  "キューの先頭要素を参照する（取り出さない）"
  [q]
  (if (queue-empty? q)
    (throw (Exception. "Queue is empty"))
    (nth (:data @q) (:front @q))))
